vulcan-core/scans/client.go

```diff
@@ -0,0 +1,185 @@
+	return &Client{coreClient, reportClient, workers}
+}
+
+// Data downloads all the reports of a given scan using, at most, the
+// specified number of goroutines in the client.
+func (c *Client) Data(ctx context.Context, ID string) (ScanData, error) {
+	scanID, err := uuid.FromString(ID)
+	if err != nil {
+		return ScanData{}, err
+	}
+
+	resp, err := c.coreC.ShowScans(ctx, coreclient.ShowScansPath(scanID))
+	if err != nil {
+		return ScanData{}, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		err = &ErrorUnexpectedStatus{http.StatusOK, resp.StatusCode}
+		return ScanData{}, fmt.Errorf("error querying scan in vulcan core %w", err)
+	}
+
+	scan, err := c.coreC.DecodeScan(resp)
+	if err != nil {
+		return ScanData{}, err
+	}
+
+	// We need the date of the scan in order to build the path for the reports
+	// of the checks belonging to it.
+
+	scanDate := scan.Scan.CreatedAt.Format("2006-01-02")
+
+	resp, err = c.coreC.ChecksScans(ctx, coreclient.ChecksScansPath(scanID))
+	if resp.StatusCode != http.StatusOK {
+		err = &ErrorUnexpectedStatus{http.StatusOK, resp.StatusCode}
+		return ScanData{}, fmt.Errorf("error querying scan checks in vulcan core %w", err)
+	}
+	checksData, err := c.coreC.DecodeChecks(resp)
+	if err != nil {
+		return ScanData{}, err
+	}
+	if checksData == nil {
+		return ScanData{}, errors.New("unexpected nil getting checks for a scan")
+	}
+
+	rs := &reportResults{}
+	rs.reports = make([]reportResult, 0, len(checksData.Checks))
+	wg := &sync.WaitGroup{}
+	checks := make(chan check, c.workers)
+
+	// Spawn a goroutine to spawn to the channel that the workers will read from.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+	LOOP:
+		for _, c := range checksData.Checks {
+			select {
+			case <-ctx.Done():
+				break LOOP
+			default:
+				checks <- check{ID: c.ID.String(), ScanID: ID, Date: scanDate}
+			}
+		}
+		close(checks)
+	}()
+
+	// Spawn the workers.
+	wg.Add(1)
+	go func() {
+		cctx, _ := context.WithCancel(ctx)
+		defer wg.Done()
+		downloadCheckReport(wg, cctx.Done(), checks, rs, c.report)
+	}()
+	wg.Wait()
+	if ctx.Err() != nil {
+		// The download process has been canceled.
+		return ScanData{}, ctx.Err()
+	}
+	var result []report.Report
+	for _, r := range rs.reports {
+		if r.err != nil {
+			return ScanData{}, r.err
+		}
+		if r.R == nil {
+			return ScanData{}, errors.New("unexpected nil downloading check report")
+		}
+		result = append(result, *r.R)
+	}
+	data := ScanData{
+		Date:    scanDate,
+		Reports: result,
+	}
+	return data, nil
+}
+
+type reportResult struct {
+	R   *report.Report
+	err error
+}
+
+type reportResults struct {
+	reports []reportResult
+	sync.Mutex
+}
+
+func (r *reportResults) Add(result reportResult) {
+	r.Lock()
+	defer r.Unlock()
+	r.reports = append(r.reports, result)
+}
+
+type check struct {
+	ID     string
+	ScanID string
+	Date   string
+}
+
+func downloadCheckReport(wg *sync.WaitGroup, done <-chan struct{}, checks <-chan check, results *reportResults, rclient ReportClient) {
```

Comment:
`wg` parameter is declared but not used inside `downloadCheckReport `, why?

Reply:
It is not needed any more, I'm will remove it.